pythnet/message_buffer/programs/message_buffer/src/instructions/put_all.rs

```diff
@@ -1,53 +1,46 @@
 use {
-    crate::{
-        state::*,
-        MessageBufferError,
-    },
+    crate::state::*,
     anchor_lang::prelude::*,
     std::mem,
 };
 
 
 pub fn put_all<'info>(
     ctx: Context<'_, '_, '_, 'info, PutAll<'info>>,
-    base_account_key: Pubkey,
+    _base_account_key: Pubkey,
     messages: Vec<Vec<u8>>,
 ) -> Result<()> {
-    let cpi_caller_auth = ctx.accounts.whitelist_verifier.is_allowed()?;
-    let message_buffer_account_info = ctx
-        .remaining_accounts
-        .first()
-        .ok_or(MessageBufferError::MessageBufferNotProvided)?;
+    ctx.accounts.whitelist_verifier.is_allowed()?;
 
-    MessageBuffer::check_discriminator(message_buffer_account_info)?;
+    let mut end_offsets = [0u16; u8::MAX as usize];
+    {
+        let msg_buffer_ai = ctx.accounts.message_buffer.to_account_info();
+        let account_data = &mut msg_buffer_ai.try_borrow_mut_data()?;
+        let header_end_index = mem::size_of::<MessageBuffer>() + 8;
 
-    let account_data = &mut message_buffer_account_info.try_borrow_mut_data()?;
-    let header_end_index = mem::size_of::<MessageBuffer>() + 8;
+        let (_, body_bytes) = account_data.split_at_mut(header_end_index);
 
-    let (header_bytes, body_bytes) = account_data.split_at_mut(header_end_index);
-
-    let message_buffer: &mut MessageBuffer = bytemuck::from_bytes_mut(&mut header_bytes[8..]);
-
-    message_buffer.validate(
-        message_buffer_account_info.key(),
-        cpi_caller_auth,
-        base_account_key,
-    )?;
-
-    message_buffer.refresh_header();
-
-    let (num_msgs, num_bytes) = message_buffer.put_all_in_buffer(body_bytes, &messages);
-
-    if num_msgs != messages.len() {
-        msg!("unable to fit all messages in MessageBuffer account. Wrote {}/{} messages and {} bytes", num_msgs, messages.len(), num_bytes);
+        let (num_msgs, num_bytes) =
+            MessageBuffer::put_all_in_buffer(body_bytes, &messages, &mut end_offsets);
```

Comment:
Seems a bit inefficient that you need to copy the end_offsets first to this buffer and then to the header.

Reply:
Might be worth zero-copy deserializing this manually because the logic feels very strange both here and in the tests after

Reply:
actually there's no need for me to switch to this. The old implementation will work. I wanted to use the load_mut() for checking the discriminator but i'm already calling `load()` in `bump = message_buffer.load()?.bump` thanks